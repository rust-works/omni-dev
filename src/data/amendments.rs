//! Amendment data structures and validation

use anyhow::{Context, Result};
use serde::{Deserialize, Serialize};
use std::fs;
use std::path::Path;

/// Amendment file structure
#[derive(Debug, Serialize, Deserialize)]
pub struct AmendmentFile {
    /// List of commit amendments to apply
    pub amendments: Vec<Amendment>,
}

/// Individual commit amendment
#[derive(Debug, Serialize, Deserialize)]
pub struct Amendment {
    /// Full 40-character SHA-1 commit hash
    pub commit: String,
    /// New commit message
    pub message: String,
}

impl AmendmentFile {
    /// Load amendments from YAML file
    pub fn load_from_file<P: AsRef<Path>>(path: P) -> Result<Self> {
        let content = fs::read_to_string(&path).with_context(|| {
            format!("Failed to read amendment file: {}", path.as_ref().display())
        })?;

        let amendment_file: AmendmentFile =
            serde_yaml::from_str(&content).context("Failed to parse YAML amendment file")?;

        amendment_file.validate()?;

        Ok(amendment_file)
    }

    /// Validate amendment file structure and content
    pub fn validate(&self) -> Result<()> {
        // Empty amendments are allowed - they indicate no changes are needed
        for (i, amendment) in self.amendments.iter().enumerate() {
            amendment
                .validate()
                .with_context(|| format!("Invalid amendment at index {}", i))?;
        }

        Ok(())
    }

    /// Save amendments to YAML file
    pub fn save_to_file<P: AsRef<Path>>(&self, path: P) -> Result<()> {
        let yaml_content =
            serde_yaml::to_string(self).context("Failed to serialize amendments to YAML")?;

        fs::write(&path, yaml_content).with_context(|| {
            format!(
                "Failed to write amendment file: {}",
                path.as_ref().display()
            )
        })?;

        Ok(())
    }
}

impl Amendment {
    /// Create a new amendment
    pub fn new(commit: String, message: String) -> Self {
        Self { commit, message }
    }

    /// Validate amendment structure
    pub fn validate(&self) -> Result<()> {
        // Validate commit hash format
        if self.commit.len() != 40 {
            anyhow::bail!(
                "Commit hash must be exactly 40 characters long, got: {}",
                self.commit.len()
            );
        }

        if !self.commit.chars().all(|c| c.is_ascii_hexdigit()) {
            anyhow::bail!("Commit hash must contain only hexadecimal characters");
        }

        if !self
            .commit
            .chars()
            .all(|c| c.is_ascii_lowercase() || c.is_ascii_digit())
        {
            anyhow::bail!("Commit hash must be lowercase");
        }

        // Validate message content
        if self.message.trim().is_empty() {
            anyhow::bail!("Commit message cannot be empty");
        }

        Ok(())
    }
}
