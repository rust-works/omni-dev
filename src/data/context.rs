//! Context data structures for enhanced commit message analysis

use anyhow::Result;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::path::PathBuf;

/// Complete context information for intelligent commit message improvement
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct CommitContext {
    /// Project-level context and conventions
    pub project: ProjectContext,
    /// Branch analysis and work pattern detection
    pub branch: BranchContext,
    /// Multi-commit analysis and work patterns
    pub range: CommitRangeContext,
    /// File-specific context and architectural understanding
    pub files: Vec<FileContext>,
    /// User-provided context information
    pub user_provided: Option<String>,
}

/// Project-level context discovered from configuration files
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct ProjectContext {
    /// Project-specific commit guidelines from .omni-dev/commit-guidelines.md
    pub commit_guidelines: Option<String>,
    /// Project-specific PR guidelines from .omni-dev/pr-guidelines.md
    pub pr_guidelines: Option<String>,
    /// Valid scopes and their descriptions from .omni-dev/scopes.yaml
    pub valid_scopes: Vec<ScopeDefinition>,
    /// Feature-specific context from .omni-dev/context/
    pub feature_contexts: HashMap<String, FeatureContext>,
    /// Parsed conventions from CONTRIBUTING.md
    pub project_conventions: ProjectConventions,
    /// Detected ecosystem (rust, node, python, etc.)
    pub ecosystem: Ecosystem,
}

/// Definition of a valid scope in the project
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ScopeDefinition {
    /// Name of the scope
    pub name: String,
    /// Human-readable description of the scope
    pub description: String,
    /// Example commit messages using this scope
    pub examples: Vec<String>,
    /// File patterns that match this scope
    pub file_patterns: Vec<String>,
}

/// Context for a specific feature or work area
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FeatureContext {
    /// Name of the feature context
    pub name: String,
    /// Description of the feature or work area
    pub description: String,
    /// Associated scope for this feature
    pub scope: String,
    /// Specific conventions for this feature
    pub conventions: Vec<String>,
}

/// Project conventions parsed from documentation
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct ProjectConventions {
    /// Expected commit message format
    pub commit_format: Option<String>,
    /// Required trailers like Signed-off-by
    pub required_trailers: Vec<String>,
    /// Preferred commit types for this project
    pub preferred_types: Vec<String>,
    /// Scope usage requirements and definitions
    pub scope_requirements: ScopeRequirements,
}

/// Requirements and validation rules for commit scopes
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct ScopeRequirements {
    /// Whether a scope is required in commit messages
    pub required: bool,
    /// List of valid scope names
    pub valid_scopes: Vec<String>,
    /// Mapping from file patterns to suggested scopes
    pub scope_mapping: HashMap<String, Vec<String>>, // file patterns -> scope
}

/// Detected project ecosystem
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub enum Ecosystem {
    #[default]
    /// Unknown or undetected ecosystem
    Unknown,
    /// Rust ecosystem with Cargo
    Rust,
    /// Node.js ecosystem with npm/yarn
    Node,
    /// Python ecosystem with pip/poetry
    Python,
    /// Go ecosystem with go modules
    Go,
    /// Java ecosystem with Maven/Gradle
    Java,
    /// Generic project without specific ecosystem
    Generic,
}

/// Branch analysis and work pattern detection
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct BranchContext {
    /// Type of work being performed on this branch
    pub work_type: WorkType,
    /// Extracted scope from branch name
    pub scope: Option<String>,
    /// Associated ticket or issue ID
    pub ticket_id: Option<String>,
    /// Human-readable description of the branch purpose
    pub description: String,
    /// Whether this is a feature branch (vs main/master)
    pub is_feature_branch: bool,
    /// Base branch this was created from
    pub base_branch: Option<String>,
}

/// Type of work being performed
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub enum WorkType {
    #[default]
    /// Unknown or unspecified work type
    Unknown,
    /// New feature development
    Feature,
    /// Bug fix
    Fix,
    /// Documentation changes
    Docs,
    /// Code refactoring
    Refactor,
    /// Maintenance tasks
    Chore,
    /// Test-related changes
    Test,
    /// CI/CD pipeline changes
    Ci,
    /// Build system changes
    Build,
    /// Performance improvements
    Perf,
}

impl std::str::FromStr for WorkType {
    type Err = anyhow::Error;

    fn from_str(s: &str) -> Result<Self> {
        match s.to_lowercase().as_str() {
            "feature" | "feat" => Ok(WorkType::Feature),
            "fix" | "bugfix" => Ok(WorkType::Fix),
            "docs" | "doc" => Ok(WorkType::Docs),
            "refactor" | "refact" => Ok(WorkType::Refactor),
            "chore" => Ok(WorkType::Chore),
            "test" | "tests" => Ok(WorkType::Test),
            "ci" => Ok(WorkType::Ci),
            "build" => Ok(WorkType::Build),
            "perf" | "performance" => Ok(WorkType::Perf),
            _ => Ok(WorkType::Unknown),
        }
    }
}

impl std::fmt::Display for WorkType {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            WorkType::Unknown => write!(f, "unknown work"),
            WorkType::Feature => write!(f, "feature development"),
            WorkType::Fix => write!(f, "bug fix"),
            WorkType::Docs => write!(f, "documentation update"),
            WorkType::Refactor => write!(f, "refactoring"),
            WorkType::Chore => write!(f, "maintenance"),
            WorkType::Test => write!(f, "testing"),
            WorkType::Ci => write!(f, "CI/CD"),
            WorkType::Build => write!(f, "build system"),
            WorkType::Perf => write!(f, "performance improvement"),
        }
    }
}

/// Multi-commit analysis and work patterns
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct CommitRangeContext {
    /// Related commit hashes in this analysis
    pub related_commits: Vec<String>, // commit hashes
    /// Files that appear in multiple commits
    pub common_files: Vec<PathBuf>,
    /// Detected work pattern across commits
    pub work_pattern: WorkPattern,
    /// Analysis of scope consistency
    pub scope_consistency: ScopeAnalysis,
    /// Overall architectural impact assessment
    pub architectural_impact: ArchitecturalImpact,
    /// Significance of changes for commit message detail
    pub change_significance: ChangeSignificance,
}

/// Detected work pattern across commits
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub enum WorkPattern {
    #[default]
    /// Unknown or undetected pattern
    Unknown,
    /// Building feature step by step
    Sequential,
    /// Multiple small cleanup commits
    Refactoring,
    /// Investigation and fixes
    BugHunt,
    /// Documentation updates
    Documentation,
    /// Config and setup changes
    Configuration,
}

/// Analysis of scope consistency across commits
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct ScopeAnalysis {
    /// Most consistent scope across commits if any
    pub consistent_scope: Option<String>,
    /// All scope changes detected
    pub scope_changes: Vec<String>,
    /// Confidence level in scope consistency (0.0-1.0)
    pub confidence: f32,
}

/// Impact on system architecture
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub enum ArchitecturalImpact {
    #[default]
    /// Small changes, no architecture impact
    Minimal,
    /// Some architectural changes
    Moderate,
    /// Major architectural changes
    Significant,
    /// Breaking changes
    Breaking,
}

/// Significance of changes for commit message detail level
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub enum ChangeSignificance {
    #[default]
    /// Simple fix or small addition
    Minor,
    /// Notable feature or improvement
    Moderate,
    /// Significant enhancement or new capability
    Major,
    /// Major system changes or breaking changes
    Critical,
}

/// File-based context and architectural understanding
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FileContext {
    /// Path to the file
    pub path: PathBuf,
    /// Purpose of this file in the project
    pub file_purpose: FilePurpose,
    /// Architectural layer this file belongs to
    pub architectural_layer: ArchitecturalLayer,
    /// Impact of changes to this file
    pub change_impact: ChangeImpact,
    /// Significance of this file in the project
    pub project_significance: ProjectSignificance,
}

/// Purpose of the file in the project
#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum FilePurpose {
    /// Configuration files
    Config,
    /// Test files
    Test,
    /// Docs and README files
    Documentation,
    /// Main application logic
    CoreLogic,
    /// API definitions, public interfaces
    Interface,
    /// Build and deployment files
    Build,
    /// Development tools and scripts
    Tooling,
}

/// Architectural layer of the file
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, Hash)]
pub enum ArchitecturalLayer {
    /// UI, CLI, web interfaces
    Presentation,
    /// Core business logic
    Business,
    /// Data access, models, storage
    Data,
    /// System, network, deployment
    Infrastructure,
    /// Cross-cutting concerns (logging, auth, etc.)
    Cross,
}

/// Impact of changes to this file
#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum ChangeImpact {
    /// Formatting, comments, style changes
    Style,
    /// New functionality, non-breaking
    Additive,
    /// Changing existing functionality
    Modification,
    /// Breaking existing functionality
    Breaking,
    /// Security, safety, or critical fixes
    Critical,
}

/// Significance of file in project
#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum ProjectSignificance {
    /// Common, everyday files
    Routine,
    /// Key files but not critical
    Important,
    /// Core files that define the project
    Critical,
}

impl CommitContext {
    /// Create a new empty context
    pub fn new() -> Self {
        Self::default()
    }

    /// Check if this context suggests a significant change needing detailed commit message
    pub fn is_significant_change(&self) -> bool {
        matches!(
            self.range.change_significance,
            ChangeSignificance::Major | ChangeSignificance::Critical
        ) || matches!(
            self.range.architectural_impact,
            ArchitecturalImpact::Significant | ArchitecturalImpact::Breaking
        ) || self.files.iter().any(|f| {
            matches!(f.project_significance, ProjectSignificance::Critical)
                || matches!(
                    f.change_impact,
                    ChangeImpact::Breaking | ChangeImpact::Critical
                )
        })
    }

    /// Get suggested commit message verbosity level
    pub fn suggested_verbosity(&self) -> VerbosityLevel {
        if self.is_significant_change() {
            VerbosityLevel::Comprehensive
        } else if matches!(self.range.change_significance, ChangeSignificance::Moderate)
            || self.files.len() > 1
            || self.files.iter().any(|f| {
                matches!(
                    f.architectural_layer,
                    ArchitecturalLayer::Presentation | ArchitecturalLayer::Business
                )
            })
        {
            // Be more generous with Detailed level for twiddle system:
            // - Moderate changes
            // - Multiple files
            // - UI/CLI or business logic changes
            VerbosityLevel::Detailed
        } else {
            VerbosityLevel::Concise
        }
    }
}

/// Suggested level of detail for commit messages
#[derive(Debug, Clone, Copy)]
pub enum VerbosityLevel {
    /// Single line, basic conventional commit
    Concise,
    /// Subject + brief body paragraph
    Detailed,
    /// Subject + detailed multi-paragraph body + lists
    Comprehensive,
}
