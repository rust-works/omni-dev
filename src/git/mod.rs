//! Git operations and repository management

use anyhow::{Context, Result};
use git2::Repository;

pub mod amendment;
pub mod commit;
pub mod remote;
pub mod repository;

pub use amendment::AmendmentHandler;
pub use commit::{CommitAnalysis, CommitInfo};
pub use remote::RemoteInfo;
pub use repository::GitRepository;

/// Check if the current directory is a git repository
pub fn check_git_repo() -> Result<()> {
    Repository::open(".").context("Not in a git repository")?;
    Ok(())
}

/// Check if working directory is clean
pub fn check_working_directory_clean() -> Result<()> {
    let repo = Repository::open(".").context("Failed to open git repository")?;

    let statuses = repo
        .statuses(None)
        .context("Failed to get repository status")?;

    if !statuses.is_empty() {
        anyhow::bail!(
            "Working directory is not clean. Please commit or stash changes before amending commit messages."
        );
    }

    Ok(())
}
