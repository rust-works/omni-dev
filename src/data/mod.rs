//! Data processing and serialization

use crate::git::{CommitInfo, RemoteInfo};
use serde::{Deserialize, Serialize};

pub mod amendments;
pub mod yaml;

pub use amendments::*;
pub use yaml::*;

/// Complete repository view output structure
#[derive(Debug, Serialize, Deserialize)]
pub struct RepositoryView {
    /// Explanation of field meanings and structure
    pub explanation: FieldExplanation,
    /// Working directory status information
    pub working_directory: WorkingDirectoryInfo,
    /// List of remote repositories and their main branches
    pub remotes: Vec<RemoteInfo>,
    /// List of analyzed commits with metadata and analysis
    pub commits: Vec<CommitInfo>,
    /// Branch information (only present when using branch commands)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub branch_info: Option<BranchInfo>,
}

/// Field explanation for the YAML output
#[derive(Debug, Serialize, Deserialize)]
pub struct FieldExplanation {
    /// Descriptive text explaining the overall structure
    pub text: String,
    /// Documentation for individual fields in the output
    pub fields: Vec<FieldDocumentation>,
}

/// Individual field documentation
#[derive(Debug, Serialize, Deserialize)]
pub struct FieldDocumentation {
    /// Name of the field being documented
    pub name: String,
    /// Descriptive text explaining what the field contains
    pub text: String,
}

/// Working directory information
#[derive(Debug, Serialize, Deserialize)]
pub struct WorkingDirectoryInfo {
    /// Whether the working directory has no changes
    pub clean: bool,
    /// List of files with uncommitted changes
    pub untracked_changes: Vec<FileStatusInfo>,
}

/// File status information for working directory
#[derive(Debug, Serialize, Deserialize)]
pub struct FileStatusInfo {
    /// Git status flags (e.g., "AM", "??", "M ")
    pub status: String,
    /// Path to the file relative to repository root
    pub file: String,
}

/// Branch information for branch-specific commands
#[derive(Debug, Serialize, Deserialize)]
pub struct BranchInfo {
    /// Current branch name
    pub branch: String,
}

impl Default for FieldExplanation {
    /// Create default field explanation
    fn default() -> Self {
        Self {
            text: "Field documentation for the YAML output format. Each entry describes the purpose and content of fields returned by the view command.".to_string(),
            fields: vec![
                FieldDocumentation {
                    name: "working_directory.clean".to_string(),
                    text: "Boolean indicating if the working directory has no uncommitted changes".to_string(),
                },
                FieldDocumentation {
                    name: "working_directory.untracked_changes".to_string(),
                    text: "Array of files with uncommitted changes, showing git status and file path".to_string(),
                },
                FieldDocumentation {
                    name: "remotes".to_string(),
                    text: "Array of git remotes with their URLs and detected main branch names".to_string(),
                },
                FieldDocumentation {
                    name: "commits[].hash".to_string(),
                    text: "Full SHA-1 hash of the commit".to_string(),
                },
                FieldDocumentation {
                    name: "commits[].author".to_string(),
                    text: "Commit author name and email address".to_string(),
                },
                FieldDocumentation {
                    name: "commits[].date".to_string(),
                    text: "Commit date in ISO format with timezone".to_string(),
                },
                FieldDocumentation {
                    name: "commits[].original_message".to_string(),
                    text: "The original commit message as written by the author".to_string(),
                },
                FieldDocumentation {
                    name: "commits[].in_main_branches".to_string(),
                    text: "Array of remote main branches that contain this commit (empty if not pushed)".to_string(),
                },
                FieldDocumentation {
                    name: "commits[].analysis.detected_type".to_string(),
                    text: "Automatically detected conventional commit type (feat, fix, docs, test, chore, etc.)".to_string(),
                },
                FieldDocumentation {
                    name: "commits[].analysis.detected_scope".to_string(),
                    text: "Automatically detected scope based on file paths (commands, config, tests, etc.)".to_string(),
                },
                FieldDocumentation {
                    name: "commits[].analysis.proposed_message".to_string(),
                    text: "AI-generated conventional commit message based on file changes".to_string(),
                },
                FieldDocumentation {
                    name: "commits[].analysis.file_changes.total_files".to_string(),
                    text: "Total number of files modified in this commit".to_string(),
                },
                FieldDocumentation {
                    name: "commits[].analysis.file_changes.files_added".to_string(),
                    text: "Number of new files added in this commit".to_string(),
                },
                FieldDocumentation {
                    name: "commits[].analysis.file_changes.files_deleted".to_string(),
                    text: "Number of files deleted in this commit".to_string(),
                },
                FieldDocumentation {
                    name: "commits[].analysis.file_changes.file_list".to_string(),
                    text: "Array of files changed with their git status (M=modified, A=added, D=deleted)".to_string(),
                },
                FieldDocumentation {
                    name: "commits[].analysis.diff_summary".to_string(),
                    text: "Git diff --stat output showing lines changed per file".to_string(),
                },
                FieldDocumentation {
                    name: "commits[].analysis.diff_content".to_string(),
                    text: "Full diff content showing line-by-line changes with added, removed, and context lines".to_string(),
                },
                FieldDocumentation {
                    name: "branch_info.branch".to_string(),
                    text: "Current branch name (only present in branch commands)".to_string(),
                },
            ],
        }
    }
}
