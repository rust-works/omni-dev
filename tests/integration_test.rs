use anyhow::Result;
use git2::{Repository, Signature};
use omni_dev::cli::git::AmendCommand;
use omni_dev::data::amendments::{Amendment, AmendmentFile};
use std::env;
use std::fs;
use std::path::PathBuf;
use tempfile::TempDir;

/// Test setup that creates a temporary git repository with test commits
struct TestRepo {
    _temp_dir: TempDir,
    repo_path: PathBuf,
    repo: Repository,
    commits: Vec<git2::Oid>,
}

impl TestRepo {
    fn new() -> Result<Self> {
        // Create temporary directory
        let temp_dir = tempfile::tempdir()?;
        let repo_path = temp_dir.path().to_path_buf();

        // Initialize git repository
        let repo = Repository::init(&repo_path)?;

        // Configure git user for commits
        let mut config = repo.config()?;
        config.set_str("user.name", "Test User")?;
        config.set_str("user.email", "test@example.com")?;

        Ok(TestRepo {
            _temp_dir: temp_dir,
            repo_path,
            repo,
            commits: Vec::new(),
        })
    }

    fn add_commit(&mut self, message: &str, content: &str) -> Result<git2::Oid> {
        // Create a test file
        let file_path = self.repo_path.join("test.txt");
        fs::write(&file_path, content)?;

        // Add file to index
        let mut index = self.repo.index()?;
        index.add_path(std::path::Path::new("test.txt"))?;
        index.write()?;

        // Create commit
        let signature = Signature::now("Test User", "test@example.com")?;
        let tree_id = index.write_tree()?;
        let tree = self.repo.find_tree(tree_id)?;

        let parent_commit = if let Some(last_commit_id) = self.commits.last() {
            Some(self.repo.find_commit(*last_commit_id)?)
        } else {
            None
        };

        let parents: Vec<&git2::Commit> = if let Some(ref parent) = parent_commit {
            vec![parent]
        } else {
            vec![]
        };

        let commit_id = self.repo.commit(
            Some("HEAD"),
            &signature,
            &signature,
            message,
            &tree,
            &parents,
        )?;

        self.commits.push(commit_id);
        Ok(commit_id)
    }

    fn get_commit_hash(&self, index: usize) -> Option<String> {
        self.commits.get(index).map(|oid| oid.to_string())
    }

    fn create_amendment_file(&self, amendments: Vec<(usize, &str)>) -> Result<PathBuf> {
        let amendment_file = AmendmentFile {
            amendments: amendments
                .iter()
                .filter_map(|(index, message)| {
                    self.get_commit_hash(*index)
                        .map(|hash| Amendment::new(hash, message.to_string()))
                })
                .collect(),
        };

        // Create the amendments file outside the repository to avoid it showing up as untracked
        let yaml_path = self.repo_path.parent().unwrap().join("amendments.yaml");
        amendment_file.save_to_file(&yaml_path)?;
        Ok(yaml_path)
    }
}

#[test]
fn test_amend_command_with_temporary_repo() -> Result<()> {
    // Set up temporary repository with test commits
    let mut test_repo = TestRepo::new()?;

    // Add some test commits
    let _commit1 = test_repo.add_commit("Initial commit", "Hello, world!")?;
    let _commit2 = test_repo.add_commit("Add feature", "Hello, world!\nNew feature added.")?;
    let _commit3 =
        test_repo.add_commit("Fix bug", "Hello, world!\nNew feature added.\nBug fixed.")?;

    println!("Created test repository at: {:?}", test_repo.repo_path);
    println!("Commits created:");
    for (i, commit_id) in test_repo.commits.iter().enumerate() {
        println!("  {}: {}", i, commit_id);
    }

    // Create amendment file to modify HEAD commit (tested and working)
    let amendments = vec![(2, "Fix critical bug in the new feature")];

    let amendment_file_path = test_repo.create_amendment_file(amendments)?;
    println!("Created amendment file at: {:?}", amendment_file_path);

    // Change to the test repository directory
    let original_dir = env::current_dir()?;
    env::set_current_dir(&test_repo.repo_path)?;

    // Test the amend command
    let result = std::panic::catch_unwind(|| {
        let amend_cmd = AmendCommand {
            yaml_file: amendment_file_path.to_string_lossy().to_string(),
        };

        println!("Testing amend command...");
        let result = amend_cmd.execute();
        println!("Amend command result: {:?}", result);
        result
    });

    // Restore original directory
    env::set_current_dir(&original_dir)?;

    match result {
        Ok(cmd_result) => {
            println!("Amend command completed: {:?}", cmd_result);

            // The implementation should now actually work
            assert!(cmd_result.is_ok(), "Amend command should succeed");

            // Verify that amendments were actually made
            env::set_current_dir(&test_repo.repo_path)?;
            let repo = Repository::open(".")?;
            let head = repo.head()?.target().unwrap();
            let commit = repo.find_commit(head)?;
            println!(
                "Current HEAD commit message after amendment: {}",
                commit.message().unwrap_or("")
            );

            // Restore directory again
            env::set_current_dir(&original_dir)?;

            // The HEAD commit message should have been amended
            let head_message = commit.message().unwrap_or("").trim();
            assert_eq!(
                head_message, "Fix critical bug in the new feature",
                "HEAD commit should have been amended with new message"
            );

            println!("✅ Test passed: Amend command successfully amended the commit message");
        }
        Err(e) => {
            println!("❌ Amend command panicked: {:?}", e);
            panic!("Amend command should not panic");
        }
    }

    Ok(())
}

#[test]
fn test_amendment_file_parsing() -> Result<()> {
    // Test that amendment file parsing works correctly
    let temp_dir = tempfile::tempdir()?;
    let yaml_path = temp_dir.path().join("test_amendments.yaml");

    // Create a test amendment file
    let test_yaml = r#"
amendments:
  - commit: "1234567890abcdef1234567890abcdef12345678"
    message: "Updated commit message 1"
  - commit: "abcdef1234567890abcdef1234567890abcdef12"
    message: "Updated commit message 2"
"#;

    fs::write(&yaml_path, test_yaml)?;

    // Test loading the amendment file
    let amendment_file = AmendmentFile::load_from_file(&yaml_path)?;
    assert_eq!(amendment_file.amendments.len(), 2);

    println!("✅ Amendment file parsing test passed");
    Ok(())
}

#[test]
fn test_amendment_validation() -> Result<()> {
    // Test amendment validation
    let temp_dir = tempfile::tempdir()?;
    let yaml_path = temp_dir.path().join("invalid_amendments.yaml");

    // Test with invalid commit hash (too short)
    let invalid_yaml = r#"
amendments:
  - commit: "12345"
    message: "Short hash should fail"
"#;

    fs::write(&yaml_path, invalid_yaml)?;

    // This should fail validation
    let result = AmendmentFile::load_from_file(&yaml_path);
    assert!(result.is_err());
    println!("✅ Amendment validation test passed - invalid hash rejected");

    Ok(())
}

#[test]
fn test_help_all_golden() -> Result<()> {
    // Capture the help-all output using the help generator directly
    use omni_dev::cli::help::HelpGenerator;

    let generator = HelpGenerator::new();
    let help_output = generator.generate_all_help()?;

    // Use insta for snapshot testing - this creates a .snap file with the expected output
    insta::assert_snapshot!("help_all_output", help_output);

    println!("✅ Golden test for help-all command completed");
    Ok(())
}
