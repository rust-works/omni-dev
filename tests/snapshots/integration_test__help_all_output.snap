---
source: tests/integration_test.rs
expression: help_output
---
omni-dev - A comprehensive development toolkit

A comprehensive development toolkit

Usage: omni-dev <COMMAND>

Commands:
  git       Git-related operations
  commands  Command template management
  help-all  Display comprehensive help for all commands
  help      Print this message or the help of the given subcommand(s)

Options:
  -h, --help     Print help
  -V, --version  Print version


================================================================================

omni-dev commands - Command template management

Command template management

Usage: commands <COMMAND>

Commands:
  generate  Generate command templates
  help      Print this message or the help of the given subcommand(s)

Options:
  -h, --help  Print help


================================================================================

omni-dev commands generate - Generate command templates

Generate command templates

Usage: generate <COMMAND>

Commands:
  commit-twiddle  Generate commit-twiddle command template
  pr-create       Generate pr-create command template
  all             Generate all command templates
  help            Print this message or the help of the given subcommand(s)

Options:
  -h, --help  Print help


================================================================================

omni-dev commands generate all - Generate all command templates

Generate all command templates

Usage: all

Options:
  -h, --help  Print help


================================================================================

omni-dev commands generate commit-twiddle - Generate commit-twiddle command template

Generate commit-twiddle command template

Usage: commit-twiddle

Options:
  -h, --help  Print help


================================================================================

omni-dev commands generate pr-create - Generate pr-create command template

Generate pr-create command template

Usage: pr-create

Options:
  -h, --help  Print help


================================================================================

omni-dev git - Git-related operations

Git-related operations

Usage: git <COMMAND>

Commands:
  commit  Commit-related operations
  branch  Branch-related operations
  help    Print this message or the help of the given subcommand(s)

Options:
  -h, --help  Print help


================================================================================

omni-dev git branch - Branch-related operations

Branch-related operations

Usage: branch <COMMAND>

Commands:
  info  Analyze branch commits and output repository information in YAML format
  help  Print this message or the help of the given subcommand(s)

Options:
  -h, --help  Print help


================================================================================

omni-dev git branch info - Analyze branch commits and output repository information in YAML format

Analyze branch commits and output repository information in YAML format

Usage: info [BASE_BRANCH]

Arguments:
  [BASE_BRANCH]  Base branch to compare against (defaults to main/master)

Options:
  -h, --help  Print help


================================================================================

omni-dev git commit - Commit-related operations

Commit-related operations

Usage: commit <COMMAND>

Commands:
  message  Commit message operations
  help     Print this message or the help of the given subcommand(s)

Options:
  -h, --help  Print help


================================================================================

omni-dev git commit message - Commit message operations

Commit message operations

Usage: message <COMMAND>

Commands:
  view   Analyze commits and output repository information in YAML format
  amend  Amend commit messages based on a YAML configuration file
  help   Print this message or the help of the given subcommand(s)

Options:
  -h, --help  Print help


================================================================================

omni-dev git commit message amend - Amend commit messages based on a YAML configuration file

Amend commit messages based on a YAML configuration file

Usage: amend <YAML_FILE>

Arguments:
  <YAML_FILE>  YAML file containing commit amendments

Options:
  -h, --help  Print help


================================================================================

omni-dev git commit message view - Analyze commits and output repository information in YAML format

Analyze commits and output repository information in YAML format

Usage: view [COMMIT_RANGE]

Arguments:
  [COMMIT_RANGE]  Commit range to analyze (e.g., HEAD~3..HEAD, abc123..def456)

Options:
  -h, --help  Print help


================================================================================

omni-dev help-all - Display comprehensive help for all commands

Display comprehensive help for all commands

Usage: help-all

Options:
  -h, --help  Print help
